L63-core-call-credentials-debug-string.md

```diff
@@ -0,0 +1,117 @@
+grpc_composite_call_credentials. They will be discussed in separate sections
+
+![C CallCredentials hierarchy](L63_graphics/c_call_creds_hierarchy.png)
+
+### GRPC Oauth2 Token fetcher credentials (`grpc_oauth2_token_fetcher_credentials`)
+Return that the token is present/absent and the token expiration time.
+
+### GRPC Google IAM credentials (`grpc_google_iam_credentials`)
+Return that the token is present/absent and the authority selector.
+
+### GRPC Service Account JWT Credentials (`grpc_service_account_jwt_access_credentials`)
+Return the token expiration date.
+
+### GRPC Access Token Credentials (`grpc_access_token_credentials`)
+Return that the token is present/absent.
+
+### GRPC Plugin Credentials (`grpc_plugin_credentials`)
+The grpc plugin credentials, as the name suggests, uses plugins to extract
+auth metadata from a different kind of credentials.
+The extraction and other actions are achieved by maintaining callbacks in the
+plugin. To support a debug string we will similarly add callbacks to return a
+debug string and then the underlying plugin will determine how to populate
+that string.
+
+To do this we need demonstrate the flow of how a callback is invoked by the
+plugin
+
+![GRPC metadata creds plugin codeflow](L63_graphics/plugin_creds_codeflow.png)
+
+* The grpc_plugin_credentials composes a grpc_metadata_credentials_plugin
+that has all the callbacks to interact with the Credential type.
+* We add a grpc_metadata_credentials::`(*debug_string)(void *state)`
+callback where state translates to the MetadataCredentialsPlugin class.
+* The MetadataCredentialsPluginWrapper provides the translation to refer to
+the C++ MetadataCredentialsPlugin class from the C Core grpc_plugin_credentials.
+We add a DebugString method here, as well as the logic of routing to the
+MetadataCredentialsPlugin’s DebugString() method. This method is then passed as
+the callback to the C Core grpc_plugin_credentials class.
+* Next step of course is to have a DebugString() method in the
+MetadataCredentialsPlugin class. The MetadataCredentialsPlugin is again a base
+class and all plugins that convert one particular type of Credentials to
+CallCredentials are derived from this class. The DebugString implementation
+returns an empty string by default.
+* Then we implement the debug string for the
+respective classes derived from MetadataCredentialsPlugin.
+
+### GRPC Composite Call Credentials (`grpc_composite_call_credentials`)
+
+The GRPC Composite call credentials maintains a list of CallCredentials.
+This means that the debug string for this struct would need to be a
+concatenation of the CallCredentials this class composes. Also since any of the
+items in the list could be a Composite Call Credential, we would need to recurse
+through them and concatenate it with the parent’s debug string. Since this could
+be unbounded and would lead to multiple memory allocations and the pain of
+concatenating and passing strings.
```

Comment:
I think this last paragraph isn't consistent with Mark's comments on the original doc. Also, the last sentence dangles.